README.md
lolologist
==========
[![Build Status](https://travis-ci.org/arusahni/lolologist.svg?branch=dev)](https://travis-ci.org/arusahni/lolologist)

A python-based tool that automatically generates an image macro from your webcam whenever you commit to a git repository. Inspired by [lolcommits](https://github.com/mroth/lolcommits).

Installing
------------

### Linux

*BEFORE* installing lolologist, make sure that the following packages are installed locally:

* python-dev
* libfreetype6-dev
* libjpeg-dev
* mplayer
* pandoc (if doing local development)

If pillow is already installed and you do not have the packages above, you need to uninstall it and install those packages. This is because when pillow is installed, it compiles optional features based on the availability of those packages. You would do this with commands:

```console
sudo pip uninstall pillow
sudo apt-get install python-dev libfreetype6-dev libjpeg-dev

# Optional, for a better font
sudo apt-get install ttf-mscorefonts-installer
```

To install lolologist, after confirming that the above prerequisites are installed, run `sudo python setup.py install`, or install via pip.

### OS X

*BEFORE* installing lolologist, make sure that the following Homebrew packages (or their equivalents) are installed locally.

* freetype
* jpeg
* imagesnap

Also, allow the locate daemon to run. This is necessary for font detection.

```console
sudo launchctl load -w /System/Library/LaunchDaemons/com.apple.locate.plist
```

To install lolologist, after confirming that the above prerequisites are installed, run `sudo python setup.py install`, or install via pip.

Using
-----
1. Within the root of your git repository, type `lolologist register`. This should add a githook that will trigger the program every time you commit.
2. Commit!

The path to your photo will be printed in the commit output.  The path is configurable - see the `Output*` options in the configuration section below.

Fonts
-----
Due to licensing concerns, I can't distribute lolologist with the iconic Impact TrueType font.  To account for this, lolologist will use your system's Impact if it exists, or fall back to an open font.

If Impact isn't installed on your system, [download and install it](http://www.fontpalace.com/font-details/Impact), and then run either `lolologist setfont` or `lolologist setfont <path-to-font>` to load it.

Configuration
-------------
The utility can be configured through the `.lolologistrc` file, usually found in your home directory. If the file doesn't exist, feel free to create it.  The following fields are accepted:

| Field             | Description                                                                  |
| ----------------- | --------------------------------------------------------------------------   |
| `Camera`          | The video device to use. (e.g. for Linux: `/dev/video1`, for OS X: `iSight`) |
| `FontPath`        | The full path to the Impact font's TTF file                                  |
| `Lolspeak`        | `on` if commit messages should be translated to lolspeak, `off` otherwise    |
| `OutputDirectory` | The format string for the directory into which all images will be placed     |
| `OutputFilename`  | The format string for the name of the generated file                         |
| `OutputFormat`    | The type of image to generate (e.g. `jpg`)                                   |
| `UploadImages`    | `on` if macros should be uploaded to the internet, `off` otherwise           |
| `UploadUrl`       | The URL to post the generated image macro to                                 |

Pythonic format strings are accepted for the outpute file name, with the caveat that *percent signs have to be escaped with another percent sign*.

For example, if you wanted to group images by the commit year and month, you could use the following:

```ini
OutputDirectory = ~/.lolologist/{project}/{time:%%Y}/{time:%%m}
```

Possible format parameters:

| Parameter  | Type       | Description                                |
| ---------- | :--------: | ------------------------------------------ |
| `project`  | *string*   | The name of the git repository's directory |
| `revision` | *string*   | A ten character ref sha                    |
| `message`  | *string*   | The entire commit message.                 |
| `time`     | *datetime* | The time of the commit.                    |

=======

Acknowledgements
----------------
Thanks go out to Stephen Newey for his [PyLOLz module](https://code.google.com/p/pylolz/).

![Analytics](https://ga-beacon.appspot.com/UA-46766795-1/lolologist/README?pixel)


CHANGELOG.md
Releases
========

v0.5.5 0 2016-06-15
-------------------
Back at it again

### Bugfixes

* Lock Pillow dependency due to a weird C bug.

v0.5.4 - 2016-06-14
-------------------
Dusting off the keyboard

### Bugfixes

* Python 3.5 support.
* Some old packages were misbehaving. Updated dependencies.
* Fixed issues where printing errors raised errors of their own.

v0.5.3 - 2015-06-06
-------------------
Hi, PyPi!

### Enhancements

* Python 3 compatibility. (#20)
* PyPy compatibility.
* Support for multiple cameras. (#23)
* Full OS X support. (#18)


v0.4.0 - 2014-06-01
-------------------
Oh hi, I didn't see you there.

### Notes
No submodule support as originally promised due to issues with the GitPython library. Will implement when their implementation is compatible with newer Git repo versions.

### Enhancements
* Removed explicit Pillow version declaration. (#17)
* Commit messages can now be translated into lolspeak.
* Images can be uploaded after generation (#15)
* OS X support w/imagesnap library (#18)

### Bugfixes
* Images are now scaled down to ~640x480 for consistency across hardware (#19)

v0.3.1 - 2013-10-13
-------------------

### Bugfixes
* Specifying a Pillow version to avoid hitting a library font sizing bug.

v0.3.0 - 2013-09-09
-------------------

### Notes
* No longer packaging Impact, and instead using LeaguGothicRegular in order to not offend the copyright gods.

### Enhancements
* User-definable fonts. (#12)

### Bugfixes
* Better repository registration handling.

v0.2.0 - 2013-09-07
-----------------
First release!


tests/test_utils.py
import sys

import pytest
import mock

from lolologist.utils import upload, LolologistError

TEST_URL = "http://test/url"
TEST_PATH = "/test/path.jpg"
TEST_PATH_BASENAME = "path.jpg"

BUILTIN_OPEN = "__builtin__.open" if sys.version_info < (3,) else "builtins.open"

class MockResponse(object):
    """ A fake requests response """
    def __init__(self, status_code=200, text=""):
        self.status_code = status_code
        self.text = text

    def json(self):
        """ Gets the response JSON """
        return {"data": {"img_url": TEST_URL + "/" + TEST_PATH_BASENAME}}

@mock.patch("requests.post")
@mock.patch("os.path.basename")
@mock.patch(BUILTIN_OPEN)
def test_upload_image_open(open_function, basename_function, post_function):
    with pytest.raises(LolologistError) as err:
        upload(TEST_URL, TEST_PATH)
    assert open_function.call_args_list[0][0][0] == TEST_PATH
    assert open_function.call_args_list[0][0][1] == "rb"
    assert "Couldn't upload the file" in err.exconly()

@mock.patch("requests.post")
@mock.patch("os.path.basename")
@mock.patch(BUILTIN_OPEN)
def test_upload_basename(open_function, basename_function, post_function):
    with pytest.raises(LolologistError) as err:
        upload(TEST_URL, TEST_PATH)
    assert basename_function.call_args_list[0][0][0] == TEST_PATH
    assert "Couldn't upload the file" in err.exconly()

@mock.patch("requests.post")
@mock.patch(BUILTIN_OPEN, return_value="opened image")
def test_upload_post_payload(open_function, post_function):
    with pytest.raises(LolologistError) as err:
        upload(TEST_URL, TEST_PATH)
    _, kwargs = post_function.call_args_list[0]
    assert 'files' in kwargs
    assert 'file' in kwargs['files']
    assert kwargs['files']['file'][0] == TEST_PATH_BASENAME
    assert kwargs['files']['file'][1] == "opened image"
    assert "Couldn't upload the file" in err.exconly()

@mock.patch("requests.post")
@mock.patch("os.path.basename")
@mock.patch(BUILTIN_OPEN, return_value="opened image")
def test_upload_post_url(open_function, basename_function, post_function):
    with pytest.raises(LolologistError) as err:
        upload(TEST_URL, TEST_PATH)
    assert post_function.call_args_list[0][0][0] == TEST_URL
    assert "Couldn't upload the file" in err.exconly()

@mock.patch("requests.post", return_value=MockResponse(status_code=200))
@mock.patch("os.path.basename")
@mock.patch(BUILTIN_OPEN, return_value="opened image")
def test_upload_success(open_function, basename_function, post_function):
    upload(TEST_URL, TEST_PATH)
    assert post_function.called

@mock.patch("requests.post", return_value=MockResponse(status_code=200))
@mock.patch("os.path.basename")
@mock.patch(BUILTIN_OPEN, return_value="opened image")
def test_upload_success_payload(open_function, basename_function, post_function):
    image_url = upload(TEST_URL, TEST_PATH)
    assert post_function.return_value.json()["data"]["img_url"] == image_url

@mock.patch("requests.post", return_value=MockResponse(status_code=500))
@mock.patch("os.path.basename")
@mock.patch(BUILTIN_OPEN, return_value="opened image")
def test_upload_failure_status(open_function, basename_function, post_function):
    with pytest.raises(LolologistError) as err:
        upload(TEST_URL, TEST_PATH)
    assert post_function.called
    assert "Couldn't upload the file" in err.exconly()
    assert post_function.return_value.status_code == 500


tests/test_cameras.py
import pytest
import mock

from lolologist.cameras import Camera

class TestBaseCamera(object):
    """Tests the base camera object functionality"""

    def test_init_warmup(self):
        c = Camera(1)
        assert c._warmup_time == 1

        c = Camera(10)
        assert c._warmup_time == 10

    def test_init_output_dir(self):
        c = Camera(1)
        assert len(c._output_directory) > 0

        c = Camera(1, '/tmp/testdir')
        assert c._output_directory == '/tmp/testdir'

    def test_init_device(self):
        c = Camera(1)
        assert c._device == None

    def test_capture(self):
        c = Camera(1)
        with pytest.raises(NotImplementedError):
            c._capture()

    @mock.patch("os.makedirs")
    def test_setup(self, makedirs_function):
        assert not makedirs_function.called
        c = Camera(1)
        c._setup()
        assert makedirs_function.called
        assert makedirs_function.call_args_list[0][0][0] == c._output_directory

    @mock.patch("lolologist.cameras.rmtree")
    @mock.patch("os.path.exists")
    def test_cleanup(self, pathexists_function, rmtree_function):
        assert not pathexists_function.called
        assert not rmtree_function.called
        c = Camera(1)
        c._cleanup()
        assert pathexists_function.called
        assert rmtree_function.called
        assert pathexists_function.call_args_list[0][0][0] == c._output_directory
        assert pathexists_function.call_args_list[0][0][0] == c._output_directory




tests/test_lolologist.py
from __future__ import unicode_literals

import pytest
import mock

from lolologist.lolologist import ImageMacro

SAMPLE_PATH = '/sample/path.jpg'
TOP_TEXT = 'This is top text'
BOTTOM_SHORT_TEXT = 'This is bottom text'
BOTTOM_WRAP_TEXT_0 = 'This is some text that should'
BOTTOM_WRAP_TEXT_1 = 'wrap to a subsequent text line'
BOTTOM_WRAP_TEXT_2 = 'unless it is too long to display'

class TestImageMacro(object):

    @mock.patch('os.path.dirname', return_value='/dir/')
    def test_init_basic(self, dirname_function):
        macro = ImageMacro(SAMPLE_PATH, TOP_TEXT, BOTTOM_SHORT_TEXT, 'font.ttf')
        assert macro.font == '/dir/font.ttf'
        assert macro.top_text == TOP_TEXT
        assert isinstance(macro.bottom_text, list)
        assert macro.bottom_text[0] == 'This is bottom text'
        assert macro.size == (0, 0)

    def test_bottom_wrapping(self):
        macro = ImageMacro(SAMPLE_PATH, TOP_TEXT, ' '.join([BOTTOM_WRAP_TEXT_0, BOTTOM_WRAP_TEXT_1]), 'font.ttf')
        assert isinstance(macro.bottom_text, list)
        assert macro.bottom_text[0] == BOTTOM_WRAP_TEXT_0
        assert macro.bottom_text[1] == BOTTOM_WRAP_TEXT_1

    def test_bottom_wrapping_max(self):
        macro = ImageMacro(SAMPLE_PATH, TOP_TEXT, ' '.join([BOTTOM_WRAP_TEXT_0, BOTTOM_WRAP_TEXT_1, BOTTOM_WRAP_TEXT_2]), 'font.ttf')
        assert isinstance(macro.bottom_text, list)
        assert macro.bottom_text[0] == BOTTOM_WRAP_TEXT_0
        assert macro.bottom_text[1] == BOTTOM_WRAP_TEXT_1
        assert macro.bottom_text[2] != BOTTOM_WRAP_TEXT_2
        assert macro.bottom_text[2][:24] == BOTTOM_WRAP_TEXT_2[:24]
        assert macro.bottom_text[2][24] == '\u2026' #ellipses




lololigist.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.

"""
lololigist - an automated image macro generator for your commits. Simply tack it onto your
	git repository's commit hook and let 'er rip!

	Aru Sahni <arusahni@gmail.com>
"""

from __future__ import unicode_literals, print_function

import argparse, os, textwrap, git, sys
from PIL import Image, ImageFont, ImageDraw
from subprocess import call, STDOUT
from contextlib import contextmanager
from shutil import rmtree

try:
	from subprocess import DEVNULL
except ImportError:
	DEVNULL = open(os.devnull, 'wb')

MAX_LINES = 3
STROKE_COLOR = (0,0,0)
TEXT_COLOR = (255, 255, 255)

POST_COMMIT_FILE = """#!/bin/sh
lololigist capture
"""

class CameraSnapper(object):
	""" A picture source """
	def __init__(self, warm_up_time=7, directory='/tmp/lololigist/'):
		""" Initializes a new webcam instance """
		self.temp_directory = directory
		self.frame_offset = warm_up_time


	@contextmanager
	def capture_photo(self):
		""" Captures a photo and provides it for further processing. """
		try:
			call(['mplayer', 'tv://', '-vo', 'jpeg:outdir={}'.format(self.temp_directory),'-frames', 
				str(self.frame_offset)], stdout=DEVNULL, stderr=STDOUT)
			yield os.path.join(self.temp_directory, '{0:08d}.jpg'.format(self.frame_offset))
		finally:
			if os.path.exists(self.temp_directory):
				rmtree(self.temp_directory)


class ImageMacro(object):
	""" An image macro """
	def __init__(self, image, top, bottom, font='impact.ttf'):
		""" Initializes the macro with a base image, two lines of text and an optional font """
		self.font = font
		self.top_text = top
		self.bottom_text = textwrap.wrap(bottom, 30)
		if len(self.bottom_text) > MAX_LINES:
			self.bottom_text[MAX_LINES - 1] = self.bottom_text[MAX_LINES - 1] + '\u2026'
		self.image_path = image


	def render(self, target_path='macro.jpg'):
		""" Renders the macro and writes it out to the target path. """
		image = Image.open(self.image_path)
		self.size = image.size
		top_font_size = 32
		bottom_font_size = 48

		top_dimensions = self.__get_text_dimensions(self.top_text, top_font_size)
		top_position = (self.size[0] - 5 - top_dimensions[0], 5)

		draw = ImageDraw.Draw(image)
		self.__draw_image(draw, self.top_text, top_font_size, top_position)
		
		lines = min(len(self.bottom_text), MAX_LINES)
		print(lines)
		for row in range(lines):
			bottom_offset = ((lines - 1 - row) * bottom_font_size) + 5
			bottom_dimensions = self.__get_text_dimensions(self.bottom_text[row], bottom_font_size)
			bottom_position = (self.size[0]/2 - bottom_dimensions[0]/2,
								self.size[1] - bottom_offset - bottom_dimensions[1])

			self.__draw_image(draw, self.bottom_text[row], bottom_font_size, bottom_position)
		
		image.save(target_path)
		return target_path


	def __draw_image(self, draw, text, font_size, position, stroke_width=3): #, bottom_font_size, bottom_dimensions, stroke_width=3):
		""" Draws the text with the given attributes to the image. """
		font = ImageFont.truetype(self.font, font_size)
		for x in range(-stroke_width, stroke_width + 1):
			for y in range(-stroke_width, stroke_width + 1):
				draw.text((position[0] + x, position[1] + y), text, STROKE_COLOR, font=font)
		draw.text(position, text, TEXT_COLOR, font=font)


	def __get_text_dimensions(self, text, font_size):
		""" Gets the measurements of text rendered at a specific font size. """
		font = ImageFont.truetype(self.font, font_size)
		return font.getsize(text)



def make_macro(sha="", message=""):
	""" Creates an image macro with the given text. """
	camera = CameraSnapper()
	with camera.capture_photo() as photo:
		macro = ImageMacro(photo, sha, message)
		print(macro.render())


def get_newest_commit(repo_path='.'):
	""" Retrieves the data for the most recent commit. """
	repo = git.Repo(repo_path)
	head_ref = repo.head.reference.commit
	return {
		"revision" : head_ref.hexsha[0:10],
		"summary" : head_ref.summary,
		"message" : head_ref.message
	}


def capture(args):
	""" Capture the most recent commit and macro it! """
	commit = get_newest_commit('.') #always capturing the current repository
	make_macro(commit['revision'], commit['summary'])


def register(args):
	""" Register lololigist with a git repo. """
	print("Attempting to register with the repository '{}'".format(args.repository))
	if not os.path.isdir(os.path.join(args.repository, '.git')):
		raise Exception("The path '{}' must contain a valid git repository".format(args.repository))
	
	hooks_dir = os.path.join(args.repository,'.git','hooks')
	if not os.path.isdir(hooks_dir):
		os.makedirs(hooks_dir)

	hook_file = os.path.join(hooks_dir, 'post-commit')
	if os.path.isfile(hook_file): #TODO: Handle multiple post-commit events in the future
		raise Exception("There is already a post-commit hook registered for this repository.")

	with open(hook_file, 'w') as script:
		script.write(POST_COMMIT_FILE)

	os.chmod(hook_file, 755)
	print("Post-commit event successfully registered. Now, get commitin'!")


def deregister(args):
	""" Remove lololigist from a git repo. """
	print("Attempting to deregister from the repository '{}'".format(args.repository))
	if not os.path.isdir(os.path.join(args.repository, '.git')):
		raise Exception("The path '{}' must contain a valid git repository".format(args.repository))
	
	hooks_dir = os.path.join(args.repository,'.git','hooks')
	hook_file = os.path.join(hooks_dir, 'post-commit')
	if not os.path.isdir(hooks_dir) or not os.path.isfile(hook_file):
		raise Exception("lololigist does not appear to be registered with this repository.")

	os.remove(hook_file) #TODO: Ensure this is actually lolologist's
	print("Post-commit event successfully deregistered. I haz a sad.")
	

if __name__ == '__main__':
	parser = argparse.ArgumentParser(description="Document your work in style!")
	subparsers = parser.add_subparsers(title="action commands")

	capture_parser = subparsers.add_parser('capture', help="Capture a snapshot and apply the most recent commit")
	capture_parser.set_defaults(func=capture)

	register_parser = subparsers.add_parser('register', help="Register lololigist with a git repository")
	register_parser.add_argument('repository', nargs='?', default='.', help="The repository to register")
	register_parser.set_defaults(func=register)

	deregister_parser = subparsers.add_parser('deregister', help="Deregister lololigist from a git repository")
	deregister_parser.add_argument('repository', nargs='?', default='.', help="The repository to deregister")
	deregister_parser.set_defaults(func=deregister)

	args = parser.parse_args();
	
	try:
		args.func(args)
	except Exception as exc:
		print("ERROR: {}".format(exc.message), file=sys.stderr)


lolologist/lolologist.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.
# pylint: disable=I0011

"""
lolologist - an automated image macro generator for your commits. Simply tack it onto your
    git repository's commit hook and let 'er rip!

    Aru Sahni <arusahni@gmail.com>
"""

from __future__ import unicode_literals, print_function

import configparser
import argparse, os, textwrap, sys, logging
from PIL import Image, ImageFont, ImageDraw
from subprocess import CalledProcessError, check_output, STDOUT

from .lolz import Tranzlator

from .cameras import MplayerCamera, ImageSnapCamera
from .utils import LolologistError, upload
from .repository import GitRepository

LOG = logging.getLogger("lolologist")

# Maximum width and height of the rendered image (in pixels). These MUST be floats.
MAX_WIDTH = 640.0
MAX_HEIGHT = 480.0

MAX_LINES = 3
STROKE_COLOR = (0, 0, 0)
TEXT_COLOR = (255, 255, 255)
FALLBACK_FONT = "LeagueGothic-Regular.otf" # Change in setup.py, too
DEFAULT_UPLOAD_URL = 'http://uploads.im/api?upload'

CURRENT_PLATFORM = 0
PLATFORM_LINUX = 1
PLATFORM_OSX = 2

POST_COMMIT_FILE = """#!/bin/sh
lolologist capture
"""

def detect_platform():
    """ Detects which platform is currently being used."""
    global CURRENT_PLATFORM
    if sys.platform.startswith("linux"):
        CURRENT_PLATFORM = PLATFORM_LINUX
    elif sys.platform == 'darwin':
        CURRENT_PLATFORM = PLATFORM_OSX
    else:
        LOG.warning("Unsupported platform detected. Assuming Linux.")
        CURRENT_PLATFORM = PLATFORM_LINUX

def is_linux():
    """Determines if the current platform is Linux
    :returns: @todo

    """
    return CURRENT_PLATFORM == PLATFORM_LINUX

def is_osx():
    """Determines if the current platform is OSX

    :returns: @todo

    """
    return CURRENT_PLATFORM == PLATFORM_OSX


class ImageMacro(object):
    """ An image macro """
    def __init__(self, image, top, bottom, font):
        """ Initializes the macro with a base image, two lines of text and an optional font """
        self.font = os.path.join(os.path.dirname(__file__), font)
        self.top_text = top
        self.bottom_text = textwrap.wrap(bottom, 30)
        if len(self.bottom_text) > MAX_LINES:
            self.bottom_text[MAX_LINES - 1] = self.bottom_text[MAX_LINES - 1] + '\u2026' #pylint: disable=W1402
        self.image_path = image
        self.size = (0, 0)

    def render(self):
        """ Returns the rendered macro. """
        image = Image.open(self.image_path)

        # If the image is bigger than desired, scale it down (maintain the aspect ratio)
        if image.size[0] > MAX_WIDTH or image.size[1] > MAX_HEIGHT:
            scaling_ratio = min(MAX_WIDTH/image.size[0], MAX_HEIGHT/image.size[1])
            image.thumbnail((scaling_ratio * image.size[0], scaling_ratio * image.size[1]), Image.ANTIALIAS)

        self.size = image.size
        top_font_size = 32
        bottom_font_size = 48

        top_dimensions = self.__get_text_dimensions(self.top_text, top_font_size)
        top_position = (self.size[0] - 5 - top_dimensions[0], 3)

        draw = ImageDraw.Draw(image)
        self.__draw_image(draw, self.top_text, top_font_size, top_position)

        lines = min(len(self.bottom_text), MAX_LINES)

        for row in range(lines):
            bottom_offset = ((lines - 1 - row) * bottom_font_size) + 15
            bottom_dimensions = self.__get_text_dimensions(self.bottom_text[row], bottom_font_size)
            bottom_position = (self.size[0]/2 - bottom_dimensions[0]/2,
                    self.size[1] - bottom_offset - bottom_dimensions[1])
            self.__draw_image(draw, self.bottom_text[row], bottom_font_size, bottom_position)

        return image

    def __draw_image(self, draw, text, font_size, position, stroke_width=3):
        """ Draws the text with the given attributes to the image. """
        font = ImageFont.truetype(self.font, font_size)
        for x_off in range(-stroke_width, stroke_width + 1):
            for y_off in range(-stroke_width, stroke_width + 1):
                draw.text((position[0] + x_off, position[1] + y_off), text, STROKE_COLOR, font=font)
        draw.text(position, text, TEXT_COLOR, font=font)

    def __get_text_dimensions(self, text, font_size):
        """ Gets the measurements of text rendered at a specific font size. """
        font = ImageFont.truetype(self.font, font_size)
        return font.getsize(text)


class Config(object): #pylint: disable=R0903
    """ Handles configuration creation and access. """
    def __init__(self, section="DEFAULT"):
        """ Creates a configuration object """
        self.config_file = os.path.expanduser('~/.lolologistrc')
        if not os.path.isfile(self.config_file):
            self.__create_config()
        config = configparser.ConfigParser()
        config.read(self.config_file)
        if section not in config:
            self.__section = "DEFAULT"
        else:
            self.__section = section
        self.__parser = config[self.__section]

    def __create_config(self):
        """ Creates the file with acceptable defaults """
        config = configparser.ConfigParser()
        config['DEFAULT'] = {
            "OutputDirectory" : os.path.join(os.path.expanduser('~'), '.lolologist', '{project}'),
            "OutputFileName" : '{revision}',
            "OutputFormat" : 'jpg',
        }

        font_path = get_impact()
        if font_path:
            config['DEFAULT']['FontPath'] = font_path

        with open(self.config_file, 'w') as config_file:
            config.write(config_file)

    def __getitem__(self, field):
        """ Gets the value of a specific field """
        return self.__parser[field]

    def __len__(self):
        """ Required for Lintin'"""
        return len(self.__parser)

    def get_font(self):
        """ Gets the configuration entry for the macro font. """
        font = self.__parser.get('FontPath', FALLBACK_FONT)
        if font == FALLBACK_FONT:
            LOG.warning("No font found. Using fallback. Run `lolologist setfont --help` for more information.")
        return font

    def get_camera(self):
        """Gets the configuration entry for the active camera device

        :returns: The configured camera. `None` if one isn't configured

        """
        return self.__parser.get("Camera")

    @property
    def lol_speak(self):
        """ Returns `True` if the lolspeak translator is enabled. """
        return self.__parser.getboolean('LolSpeak', False)

    def update_config(self, setting, value):
        """ Sets a value for the specific setting in the DEFAULT section. """
        config = configparser.ConfigParser()
        config.read(self.config_file)
        config["DEFAULT"][setting] = value

        with open(self.config_file, 'w') as config_file:
            config.write(config_file)

        self.__parser = config[self.__section]

    def clear_setting(self, setting):
        """ Removes a setting from the DEFAULT section. """
        config = configparser.ConfigParser()
        config.read(self.config_file)
        config.remove_option('DEFAULT', setting)

        with open(self.config_file, 'w') as config_file:
            config.write(config_file)

        self.__parser = config[self.__section]

    @property
    def upload(self):
        """ Determines if the uploader should be triggered. """
        return self.__parser.getboolean('UploadImages', False) and len(self.upload_url) > 0

    @property
    def upload_url(self):
        """ The URL to upload to. """
        return self.__parser.get('UploadUrl', DEFAULT_UPLOAD_URL)


class Lolologist(object):
    """ The main application """

    def __init__(self, repo_path='.'):
        """Initializes a Lologogis instance at the given path
        """
        detect_platform()
        self.config = Config()
        self.repo_path = repo_path

    def __make_macro(self, revision, summary, **kwargs):
        """ Creates an image macro with the given text.

       :param revision: The SHA-1 hash of the commit
       :param summary: The short description of the commit
       :returns: The full path to the saved image

        """
        if is_osx():
            camera = ImageSnapCamera(device=self.config.get_camera())
        else:
            camera = MplayerCamera(device=self.config.get_camera())
        with camera.capture_photo() as photo:
            macro = ImageMacro(photo, revision, summary, self.config.get_font())
            image = macro.render()
            directory_path = self.config['OutputDirectory'].format(revision=revision, **kwargs)
            if not os.path.isdir(directory_path):
                os.makedirs(directory_path)
            file_path = os.path.join(self.config['OutputDirectory'], self.config['OutputFileName']).format(
                revision=revision,
                **kwargs
            ) + '.' + self.config["OutputFormat"]
            image.save(file_path)
            return file_path

    def __get_newest_commit(self):
        """ Retrieves the data for the most recent commit. """
        loltranz = Tranzlator()
        translator = loltranz.translate_sentence if self.config.lol_speak else None
        return GitRepository(self.repo_path).get_newest_commit(translator=translator)

    def capture(self, args): #pylint: disable=W0613
        """ Capture the most recent commit and macro it! """
        commit = self.__get_newest_commit()
        image = self.__make_macro(**commit)
        if self.config.upload:
            url = upload(self.config.upload_url, image)
            print("Uploaded:", url)
        print("Macro saved:", image)

    @staticmethod
    def register(args): #pylint: disable=W0613
        """ Register lolologist with a git repo. """
        print("Attempting to register with the repository '{}'".format(args.repository))
        GitRepository(args.repository).register(POST_COMMIT_FILE)

    @staticmethod
    def deregister(args): #pylint: disable=W0613
        """ Remove lolologist from a git repo. """
        print("Attempting to deregister from the repository '{}'".format(args.repository))
        GitRepository(args.repository).deregister()
        print("Post-commit event successfully deregistered. I haz a sad.")

    def set_font(self, args):
        """ Sets the default image macro font. """
        font_path = args.font_path if args.font_path else get_impact()
        if font_path:
            self.config.update_config("FontPath", font_path)
            print("Font successfully set to '{}'".format(font_path))
        else:
            raise LolologistError("Could not find Impact. Falling back to League Gothic.")

    def speak_lolz(self, args):
        """ Sets the state of lolspeak for commits """
        enabled = args.lol_on.lower() == "on" if args.lol_on else False
        if (not self.config.lol_speak and args.lol_on.lower() == "on") or \
                (self.config.lol_speak and args.lol_on.lower() == "off"):
            self.config.update_config("LolSpeak", args.lol_on.lower())
        print("Lolspeak {}!".format("enabled" if enabled else "disabled"))

    def set_uploader(self, args):
        """ Sets the state of the uploader, as well as the URL endpoint."""
        action = args.action.lower()
        if action == "on" or action == "off":
            self.config.update_config("UploadImages", action)
        if action == "clear":
            self.config.clear_setting("UploadUrl")
        elif args.url:
            self.config.update_config("UploadUrl", args.url)


def get_impact_locations():
    """ Gets allthe locations of the Impact font for the current operating system

    :returns: A collection of all the Impact locations

    """
    if is_osx():
        return check_output(['locate', '-i', 'Impact.ttf'], stderr=STDOUT)
    return check_output(['locate', '--regex', '-i', 'Impact.ttf$'], stderr=STDOUT)

def get_impact():
    """ Finds Impact.ttf on one's system

    :returns: The full path to the Impact font

    """
    font_path = None
    try:
        locs = get_impact_locations()
        for loc in locs.splitlines():
            if isinstance(loc, bytes):
                loc = loc.decode("utf-8")
            if loc.startswith("/usr") or loc.startswith("/Library"): #most likely a permanent font
                font_path = loc
                break
            elif font_path is None:
                font_path = loc
    except CalledProcessError:
        pass
    return font_path

def parse_args(app):
    """Gets the arguments passed to lolologist

    :param app: The core application object
    :returns: The parsed commands

    """
    parser = argparse.ArgumentParser(description="Document your work in style!")
    subparsers = parser.add_subparsers(title="action commands")

    capture_parser = subparsers.add_parser('capture', help="Capture a snapshot and apply the most recent commit")
    capture_parser.set_defaults(func=app.capture)

    register_parser = subparsers.add_parser('register', help="Register lolologist with a git repository")
    register_parser.add_argument('repository', nargs='?', default='.', help="The repository to register")
    register_parser.set_defaults(func=Lolologist.register)

    deregister_parser = subparsers.add_parser('deregister', help="Deregister lolologist from a git repository")
    deregister_parser.add_argument('repository', nargs='?', default='.', help="The repository to deregister")
    deregister_parser.set_defaults(func=Lolologist.deregister)

    setfont_parser = subparsers.add_parser('setfont', help="Set the font to use for image macros.")
    setfont_parser.add_argument('font_path', nargs="?",
            help="The full path to the desired font. If none is specified, attempt to find the system's Impact font."
    )
    setfont_parser.set_defaults(func=app.set_font)

    tranzlator_parser = subparsers.add_parser('speaklolz',
            help="Translate commit summaries and messages to lolzspeak.")
    tranzlator_parser.add_argument('lol_on', help="'on' to enable the translator, 'off' to disable it.")
    tranzlator_parser.set_defaults(func=app.speak_lolz)

    uploader_parser = subparsers.add_parser('uploader', help="Set file uploading preferences.")
    uploader_parser.add_argument('action', choices=['on', 'off', 'set', 'clear'],
            help="'on' to enable the uploader, 'off' to disable it, 'set' to set the url, 'clear' to clear the URL.")
    uploader_parser.add_argument('url', nargs='?', default=None, help="The URL to POST the image to.")
    uploader_parser.set_defaults(func=app.set_uploader)

    return parser.parse_args()

def main():
    """ Entry point for the application """
    app = Lolologist()
    try:
        args = parse_args(app)
        args.func(args)
    except LolologistError as exc:
        print("ERROR: {}".format(exc.message), file=sys.stderr)

if __name__ == '__main__':
    main()
